src/types/utils.cpp

```diff
@@ -378,71 +101,227 @@ til::color Utils::ColorFromHexString(const std::string_view str)
+
+                parameterValue *= 16;
+                unsigned int intVal = 0;
+                const auto ret = HexToUint(wch, intVal);
+                if (!ret)
+                {
+                    // Encountered something weird oh no
+                    return std::nullopt;
+                }
+
+                parameterValue += intVal;
+
+                if (isSharpSignFormat)
+                {
+                    // If we get this far, any number can be seen as a valid part
+                    // of this component.
+                    foundColor = true;
+
+                    if (i >= rgbHexDigitCount)
+                    {
+                        // Successfully parsed this component. Start the next one.
+                        break;
+                    }
+                }
+                else
+                {
+                    // Record the hex digit count of the current component.
+                    rgbHexDigitCount = i + 1;
+
+                    // If this is the first 2 component...
+                    if (component < 2 && curr < string.cend() && *curr == L'/')
+                    {
+                        // ...and we have successfully parsed this component, we need
+                        // to skip the delimiter before starting the next one.
+                        curr++;
+                        foundColor = true;
+                        break;
+                    }
+                    // Or we have reached the end of the string...
+                    else if (curr >= string.cend())
+                    {
+                        // ...meaning that this is the last component. We're not going to
+                        // see any delimiter. We can just break out.
+                        foundColor = true;
+                        break;
+                    }
+                }
+            }
+
+            if (!foundColor)
+            {
+                // Indicates there was some error parsing color.
+                return std::nullopt;
+            }
+
+            // Calculate the actual color value based on the hex digit count.
+            auto& colorValue = til::at(colorValues, component);
+            const auto scaleMultiplier = isSharpSignFormat ? 0x10 : 0x11;
+            const auto scaleDivisor = scaleMultiplier << 8 >> 4 * (4 - rgbHexDigitCount);
+            colorValue = parameterValue * scaleMultiplier / scaleDivisor;
+        }
+
+        if (curr >= string.cend())
+        {
+            // We're at the end of the string and we have successfully parsed the color.
+            foundValidColorSpec = true;
+        }
+    }
+
+    // Only if we find a valid colorspec can we pass it out successfully.
+    if (foundValidColorSpec)
+    {
+        return til::color(LOBYTE(colorValues.at(0)),
+                          LOBYTE(colorValues.at(1)),
+                          LOBYTE(colorValues.at(2)));
+    }
+
+    return std::nullopt;
 }
+CATCH_FAIL_FAST()

```

Comment:
I can no longer use `CATCH_RETURN_FALSE` because `false` is actually a `std::optional` that has value. 

And is it OK to use `std::optional`? Personally I love it. Just that everything else in this file is using `THROW_IF_FAILED`.

Reply:
You know, I'm totally okay with using std::optional in more places

Reply:
I am too. @miniksa?

Reply:
**HOWEVER**, I would rather not failfast.

```suggestion
catch (...)
{
    LOG_CAUGHT_EXCEPTION();
    return std::nullopt;
}
```

a VT application should not be able to take down the entire terminal